### PR TITLE
WIP

### PR DESCRIPTION
**Pull request by cowboy**
To test:

* `/dm @testbot-cowboy expertise update ruby`
* `/dm @testbot-cowboy expertise update missing`

### CODE REVIEW

File: src/robocoup/commands/expertise.js [hunk 1]
```diff
@@ -1,20 +1,35 @@
 import Promise from 'bluebird';
 import heredoc from 'heredoc-tag';
+import moment from 'moment';
 
 import {parseArgs} from '../../lib/args';
 import {query} from '../../lib/db';
 import Dialog from '../../lib/dialog';
 
+// https://github.com/bocoup/bocoup-meta/issues/243#issuecomment-187268880
+const EXPERIENCE = [
```
Comment (tkellen) [inline src/robocoup/commands/expertise.js]: Can you source this from the `experience_scale` table please?

File: src/robocoup/commands/expertise.js [hunk 2]
```diff
@@ -1,20 +1,35 @@
...
+  'I have no experience with this.',
+  'I have some experience but I am not yet confident with this.',
+  'I have some experience and feel confident with this.',
+  'I have lots of experience and can teach others this.',
+  'I would feel comfortable having a team of people rely on me for this.',
+];
+const INTEREST = [
```
Comment (tkellen) [inline src/robocoup/commands/expertise.js]: Can you source this from the `interest_scale` table please?

File: src/robocoup/commands/expertise.js [hunk 3]
```diff
@@ -324,9 +254,156 @@ addCommand('find', {
...
+    const updatePromise = query('update_expertise', user, id, experience, interest, reason || '');
+    return Promise.props({
+      oldValues,
+      // We need to wait for the update to resolve, but do we care about the result?
+      updatePromise,
+    });
+  })
+  .then(({oldValues}) => {
+    // Show a summary of the changes.
+    const summary = intExpProps.map(prop => {
```
Comment (tkellen) [inline src/robocoup/commands/expertise.js]: Consider extracting this into a named function that could be tested?

File: src/lib/dialog.js [hunk 4]
```diff
@@ -51,49 +53,148 @@ export default class Dialog {
...
+    })
+    .then(({question, result}) => {
+      // There's no question. If it's because the question was empty or a
+      // question promise resolved to nothing, get the next question. If it's
+      // because there were no more questions, just return the result.
+      if (!question) {
+        return questions.length === 0 ? result : next(result);
+      }
+      // Guess which method to call, based on the shape of the question object.
+      let apiMethod, responseMethod;
```
Comment (tkellen) [inline src/lib/dialog.js]: make these a const with a ternary assignment?

File: src/lib/dialog.js [hunk 5]
```diff
@@ -51,49 +53,148 @@ export default class Dialog {
...
+    question = `Choose one of the following:`,
+    prompt = null,
+    exit = null,
     onMatch,
+    onError = text => `_Sorry, but \`${text}\` is not a valid response. Please try again._`,
     oneTimeHeader,
   }) {
-    const keys = Object.keys(choices);
-    const message = context => [
+    let keys;
```
Comment (tkellen) [inline src/lib/dialog.js]:
```
const arr = ['cat','dog','cow'];
R.zipObj(R.range(1,arr.length+1), arr);
```

Comment (tkellen) [inline src/lib/dialog.js]: or
```
const arr = ['cat','dog','cow'];
arr.reduce(function (result, item, idx) { 
  result[idx+1] = item;
  return result;
}, {});
```

File: src/robocoup/commands/expertise.js [hunk 6]
```diff
@@ -324,9 +254,156 @@ addCommand('find', {
...
+      if (missing.length === 0) {
+        return heredoc.trim.unindent`
+          You have no outstanding expertise data.
+          View your expertise list with \`expertise me\`.
+        `;
+      }
+      const expertise = missing[0];
+      const num = missing.length === 1 ? '' : ` ${missing.length}`;
+      const oneTimeHeader = [
+        ...(header ? [header, ''] : []),
```
Comment (tkellen) [inline src/robocoup/commands/expertise.js]: a comment here about what is happening would likely help a future reader

File: src/lib/dialog.js (hunk 4 above)
Comment (cowboy) [inline src/lib/dialog.js]: If these were `const` with `?:`, the condition would need to be duplicated, which isn't ideal. `if/else` might feel ugly, but makes more sense here.

Comment (tkellen) [inline src/lib/dialog.js]: In these scenarios I personally use this opportunity to define a named thing that describes the conditional. e.g.
```
const questionHasChoices = question.choices
const responseMethod = questionHasChoices ? 'onMatch' : 'onResponse';
```

Comment (tkellen) [inline src/lib/dialog.js]: But, this is well into the realm of personal preference. Do whatcha think!

File: src/robocoup/commands/expertise.js (hunk 6 above)
Comment (cowboy) [inline src/robocoup/commands/expertise.js]: Yeah, this code is insane. I think I'm going to need to make a full commenting pass after this is all done, but I'm not sure I want to hold up merging and deploying this code to do so. If I feel pressed for time, I'll merge this as-is and create an issue to comment more thoroughly.

File: src/lib/dialog.js (hunk 4 above)
Comment (cowboy) [inline src/lib/dialog.js]: In this case, it's just an `if/else` but i anticipate the logic growing more complex as more abstractions are added.

**Comment by cowboy** [comment]
Closing in favor of mega-PR #67

**Pull request closed without merging**